File: python/src/aoc/year2022/day2.py
from aoc.util import load_input, load_example
from enum import Enum


class Shape(Enum):
    ROCK = 1
    PAPER = 2
    SCISSOR = 3


class Score(Enum):
    WIN = 6
    DRAW = 3
    LOST = 0


# opponent, me, score
result = [
    (Shape.ROCK, Shape.ROCK, Score.DRAW),
    (Shape.ROCK, Shape.PAPER, Score.WIN),
    (Shape.ROCK, Shape.SCISSOR, Score.LOST),
    (Shape.PAPER, Shape.ROCK, Score.LOST),
    (Shape.PAPER, Shape.PAPER, Score.DRAW),
    (Shape.PAPER, Shape.SCISSOR, Score.WIN),
    (Shape.SCISSOR, Shape.ROCK, Score.WIN),
    (Shape.SCISSOR, Shape.PAPER, Score.LOST),
    (Shape.SCISSOR, Shape.SCISSOR, Score.DRAW),
]

mappingOpponent = {
    "A": Shape.ROCK,
    "B": Shape.PAPER,
    "C": Shape.SCISSOR,
}

mapping1 = {
    "X": Shape.ROCK,
    "Y": Shape.PAPER,
    "Z": Shape.SCISSOR,
}

mapping2 = {
    "X": Score.LOST,
    "Y": Score.DRAW,
    "Z": Score.WIN,
}


def scoring_part1(o, me):
    m: Shape = mapping1[me]
    s: Score = next(filter(lambda t: t[0] == o and t[1] == m, result))[2]
    return m, s


def scoring_part2(o, me):
    s: Score = mapping2[me]
    m: Shape = next(filter(lambda t: t[0] == o and t[2] == s, result))[1]
    return m, s


def score_lines(lines, fun):
    for line in lines:
        if not line:
            continue
        opponent, me = line.split()
        m, r = fun(mappingOpponent[opponent], me)
        yield m.value + r.value


def part1(lines):
    """
    >>> part1(load_example(__file__, "2"))
    15
    """
    return sum(score_lines(lines, scoring_part1))


def part2(lines):
    """
    >>> part2(load_example(__file__, "2"))
    12
    """
    return sum(score_lines(lines, scoring_part2))


if __name__ == "__main__":
    data = load_input(__file__, 2022, "2")
    print(part1(data))
    print(part2(data))

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2022/Day2.kt
package io.github.ocirne.aoc.year2022

import io.github.ocirne.aoc.AocChallenge

typealias ScoringFunction = (Day2.Shape, Char) -> Pair<Day2.Shape, Day2.Score>

class Day2(val lines: List<String>) : AocChallenge(2022, 2) {

    enum class Shape(val score: Int) {
        ROCK(1),
        PAPER(2),
        SCISSOR(3),
    }

    enum class Score(val score: Int) {
        WIN(6),
        DRAW(3),
        LOST(0)
    }

    // opponent, me, score
    private val result = listOf(
        Triple(Shape.ROCK, Shape.ROCK, Score.DRAW),
        Triple(Shape.ROCK, Shape.PAPER, Score.WIN),
        Triple(Shape.ROCK, Shape.SCISSOR, Score.LOST),
        Triple(Shape.PAPER, Shape.ROCK, Score.LOST),
        Triple(Shape.PAPER, Shape.PAPER, Score.DRAW),
        Triple(Shape.PAPER, Shape.SCISSOR, Score.WIN),
        Triple(Shape.SCISSOR, Shape.ROCK, Score.WIN),
        Triple(Shape.SCISSOR, Shape.PAPER, Score.LOST),
        Triple(Shape.SCISSOR, Shape.SCISSOR, Score.DRAW),
    )

    private val mappingOpponent = mapOf(
        'A' to Shape.ROCK,
        'B' to Shape.PAPER,
        'C' to Shape.SCISSOR,
    )

    private val mapping1 = mapOf(
        'X' to Shape.ROCK,
        'Y' to Shape.PAPER,
        'Z' to Shape.SCISSOR,
    )

    private val mapping2 = mapOf(
        'X' to Score.LOST,
        'Y' to Score.DRAW,
        'Z' to Score.WIN,
    )

    private val scoringPart1: ScoringFunction = { opponentShape, me ->
        val myShape: Shape = mapping1[me]!!
        val score: Score = result.first { (o, m, _) -> o == opponentShape && m == myShape }.third
        Pair(myShape, score)
    }

    private val scoringPart2: ScoringFunction = { opponentShape, me ->
        val score: Score = mapping2[me]!!
        val myShape: Shape = result.first { (o, _, s) -> o == opponentShape && s == score }.second
        Pair(myShape, score)
    }

    private fun scoreLines(scoringFun: ScoringFunction): Int {
        return lines
            .asSequence()
            .filter { line -> line.isNotBlank() }
            .map { line -> line.trim().split(' ') }
            .map { p -> Pair(p[0].first(), p[1].first())}
            .map { (opponent, me) -> scoringFun.invoke(mappingOpponent[opponent]!!, me) }
            .map { (m, s) -> m.score + s.score }
            .sum()
    }

    override fun part1(): Int {
        return scoreLines(scoringPart1)
    }

    override fun part2(): Int {
        return scoreLines(scoringPart2)
    }
}
